service/history/workflow/activity.go

```diff
@@ -76,3 +82,89 @@ func updateActivityInfoForRetries(
 
 	return ai
 }
+
+func ActivityState(ai *persistence.ActivityInfo) enumspb.PendingActivityState {
+	if ai.CancelRequested {
+		return enumspb.PENDING_ACTIVITY_STATE_CANCEL_REQUESTED
+	}
+	if ai.StartedEventId != common.EmptyEventID {
+		return enumspb.PENDING_ACTIVITY_STATE_STARTED
+	}
+	return enumspb.PENDING_ACTIVITY_STATE_SCHEDULED
+}
+
+func GetPendingActivityInfo(
+	ctx context.Context, // only used as a passthrough to GetActivityType
+	shardContext shard.Context,
+	mutableState MutableState,
+	ai *persistence.ActivityInfo,
+) (*workflowpb.PendingActivityInfo, error) {
+	now := shardContext.GetTimeSource().Now().UTC()
+
+	p := &workflowpb.PendingActivityInfo{
+		ActivityId: ai.ActivityId,
+	}
+	if ai.GetUseWorkflowBuildIdInfo() != nil {
+		p.AssignedBuildId = &workflowpb.PendingActivityInfo_UseWorkflowBuildId{UseWorkflowBuildId: &emptypb.Empty{}}
+	} else if ai.GetLastIndependentlyAssignedBuildId() != "" {
+		p.AssignedBuildId = &workflowpb.PendingActivityInfo_LastIndependentlyAssignedBuildId{
+			LastIndependentlyAssignedBuildId: ai.GetLastIndependentlyAssignedBuildId(),
+		}
+	}
+
+	p.State = ActivityState(ai)
+
+	p.LastAttemptCompleteTime = ai.LastAttemptCompleteTime
+	if !ai.HasRetryPolicy {
+		p.Attempt = 1
+		p.NextAttemptScheduleTime = nil
+	} else {
+		p.Attempt = ai.Attempt
+		if p.State == enumspb.PENDING_ACTIVITY_STATE_SCHEDULED {
+			if now.Before(ai.ScheduledTime.AsTime()) {
+				// in this case activity is waiting for a retry
+				p.NextAttemptScheduleTime = ai.ScheduledTime
+				currentRetryDuration := p.NextAttemptScheduleTime.AsTime().Sub(p.LastAttemptCompleteTime.AsTime())
+				p.CurrentRetryInterval = durationpb.New(currentRetryDuration)
+			} else {
+				// in this case activity is at least scheduled
+				p.NextAttemptScheduleTime = nil
+				interval := ExponentialBackoffAlgorithm(ai.RetryInitialInterval, ai.RetryBackoffCoefficient, p.Attempt)
+				p.CurrentRetryInterval = durationpb.New(interval)
+			}
+		}
+		if p.Attempt < 1 {
+			p.Attempt = 1
+		}
+	}
+
+	if ai.LastHeartbeatUpdateTime != nil && !ai.LastHeartbeatUpdateTime.AsTime().IsZero() {
+		p.LastHeartbeatTime = ai.LastHeartbeatUpdateTime
+		p.HeartbeatDetails = ai.LastHeartbeatDetails
+	}
+	var err error
+	p.ActivityType, err = mutableState.GetActivityType(ctx, ai)
+	if err != nil {
+		return nil, err
+	}
+	if p.State == enumspb.PENDING_ACTIVITY_STATE_SCHEDULED {
+		p.ScheduledTime = ai.ScheduledTime
+	} else {
+		p.LastStartedTime = ai.StartedTime
+	}
+	p.LastWorkerIdentity = ai.StartedIdentity
+	if ai.HasRetryPolicy {
+		p.ExpirationTime = ai.RetryExpirationTime
+		if ai.RetryMaximumAttempts != 0 {
+			p.MaximumAttempts = ai.RetryMaximumAttempts
+		}
+		if ai.RetryLastFailure != nil {
+			p.LastFailure = ai.RetryLastFailure
+		}
+		if p.LastWorkerIdentity == "" && ai.RetryLastWorkerIdentity != "" {
+			p.LastWorkerIdentity = ai.RetryLastWorkerIdentity
+		}
```

Comment:
When would `p.LastWorkerIdentity` not be `== ""`?

```suggestion
		p.LastWorkerIdentity = ai.RetryLastWorkerIdentity
```

Reply:
I have no clue, this is the original code.

Reply:
actually few lines above:
p.LastWorkerIdentity = ai.StartedIdentity